server/src/main/java/org/opensearch/cluster/coordination/JoinTaskExecutor.java

```diff
@@ -422,6 +439,45 @@ public static void ensureNodeCommissioned(DiscoveryNode node, Metadata metadata)
         }
     }
 
+    /**
+     * The method ensures two conditions -
+     * 1. The joining node is remote store if it is joining a remote store cluster.
+     * 2. The joining node is non-remote store if it is joining a non-remote store cluster.
+     * A remote store node is the one which holds the all the remote store attributes and a remote store cluster is
+     * the one which has only homogeneous remote store nodes with same node attributes
+     *
+     * @param joiningNode
+     * @param currentState
+     */
+    public static void ensureRemoteStoreNodesCompatibility(DiscoveryNode joiningNode, ClusterState currentState) {
+        List<DiscoveryNode> existingNodes = new ArrayList<>(currentState.getNodes().getNodes().values());
+
+        /**
+         * If there are no node in the cluster state we will No op the compatibility check as at this point we
+         * cannot determine if this is a remote store cluster or non-remote store cluster.
+         */
+        if (existingNodes.size() == 0) {
+            return;
+        }
+
+        /**
+         * TODO: The below check is valid till we support migration, once we start supporting migration a remote
+         * store node will be able to join a non remote store cluster and vice versa. #7986
+         */
+        if (isRemoteStoreNode(joiningNode)) {
```

Comment:
nit - we can simplify the if else logic here by determining remote status of new node and existing node upfront and making this if else a  simple condition. 

```
boolean newNodeRemoteStatus = isRemoteStoreNode(joiningNode)
boolean existingNodeRemoteStatus = isRemoteStoreNode(existingNodes.get(0))
boolean sameAttributes = compareNodeAttributes(joiningNode, existingNode) // let this return boolean
if ((newNodeRemoteStatus != existingNodeRemoteStatus) || (newNodeRemoteStatus && sameAttributes == false)) {
   // throw..
}
```